### ISSUE TITLE
Performance: [SQS] Sending an SQS message while broadcasting SNS messages can be very slow

### ISSUE THREAD

**Issue by davy-oo** [localstack/localstack#7101]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

The attached test case  [TestSQS-LoadImpactOnSendMessage.zip](https://github.com/localstack/localstack/files/9878986/TestSQS-LoadImpactOnSendMessage.zip) uses 3 services defined in a docker-compose.yml:
- `broadcastPublisher` service which is responsible for publishing messages to SNS topic.
- `broadcastSubscriber` service which subscribes to the above SNS topic and reads and deletes all received messages from its SQS queue.
- `sqs` service which sends messages to normal SQS queue

 During the test, SQS messages are sent while the broadcastPublisher and broadcastSubscriber generate load in LocalStack. The call done by sqs service("sqsClient.SendMessageAsync(queue.QueueUrl, message)") can take a very long time to complete the more subscribers are added.

For example, sending 200 SQS messages without any broadcast in series takes about 3 seconds. If 5 broadcastSubscriber services are generating traffic(broadcastPublisher constantly publishing messages to SNS topic), sending the 200 SQS messages would take about 40s, so 10 times slower.

The same tests running against AWS do not show slowness.

See below a table that shows the slow down when adding more and more subscribers:

| Subscribers count | Time to send 200 messages [s] | Min time to send a single message [ms] | Max time to send single message [ms]
|:-----------------:|:-----------------------------:|:--------------------------------------:|:-----------------------------------:|
|  0                | 3.00                     | 10                                | 64                           |
|  1                | 12.21                     | 17                                | 404                            |
|  2                | 24.26                    | 29                               | 329                             |
|  3                | 25.61                    | 33                              | 388                            |
|  4                | 32.25                    | 40                               | 415                            |
|  5                | 40.00                     | 52                              | 482                           |
|  6                | 46.90                    | 45                                | 560                           |
|  7                | 53.83                    | 72                               | 638                          |

In our real pipelines, we test about 10 services together, sending an SQS message can take almost 1 second if broadcast messages are processed.

### Expected Behavior

Testing a bunch of micro-services using AWS SQS/SNS messaging capability should be possible with localstack without being too slow. At least not by a magnitude of 10 times.

### How are you starting LocalStack?

With a `docker run` command

#### Create a docker network and Start LocalStack
Docker for Desktop is configured with Linux containers.
```bash
  docker network create linux-dependencies
  docker run --network linux-dependencies --name localstack_linux -p 4566:4566 -e SERVICES=sqs,sns,logs -e EAGER_SERVICE_LOADING=1 localstack/localstack:1.2.0
```

#### Prepare test case

1. Download C# .Net6 attached test case [TestSQS-LoadImpactOnSendMessage.zip](https://github.com/localstack/localstack/files/9878986/TestSQS-LoadImpactOnSendMessage.zip)
2. Extract and review readme.md
3. Build docker images: 
  - Configure Docker for Desktop to run linux containers
  - Change to directory where dockerfile files are.
  - Build the 3 images:
    - `docker build -t testsqs-broadcast-publisher -f Dockerfile-BroadcastPublisher .`
    - `docker build -t testsqs-broadcast-subscriber -f Dockerfile-BroadcastSubscriber .`
    - `docker build -t testsqs-sqs -f Dockerfile-SQS .`

#### Run the test using docker compose
The test can be run with various parameters:
- `docker compose up --scale broadcastSubscriber=7` - runs single instance for `sqs` and `broadcastPublisher` and 7 instances for `broadcastSubscriber`, this is the slowest
- `docker compose up` - runs single instance for every service
- `docker compose up --scale broadcastPublisher=0 --scale broadcastSubscriber=0` - runs only `sqs` service, this is the fastest

Review the output of the sqs container for results: `docker compose logs --tail 20 sqs`

For example:
```
testsqs-loadimpactonsendmessage-sqs-1  | ==> 10:28:07 | TestSQS-SQS | SQS publishing for message number 198 took 105.6228.
testsqs-loadimpactonsendmessage-sqs-1  | ==> 10:28:08 | TestSQS-SQS | SQS publishing for message number 199 took 109.9607.
testsqs-loadimpactonsendmessage-sqs-1  | ==> 10:28:08 | TestSQS-SQS | 200 SQS messages published in 53.8309176 [s].
testsqs-loadimpactonsendmessage-sqs-1  | ==> 10:28:08 | TestSQS-SQS | Min time to send single message 72.4197 [ms].
testsqs-loadimpactonsendmessage-sqs-1  | ==> 10:28:08 | TestSQS-SQS | Max time to send single message 638.3528 [ms].
testsqs-loadimpactonsendmessage-sqs-1  | ==> 10:28:08 | TestSQS-SQS | Finished.
```

### Environment

```markdown
- OS: Windows 10
- LocalStack: 1.2.0
- Docker For Desktop 4.12.0 (85629), configured for Linux containers, Engine: 20.10.17
```

**Comment by bentsku** [comment]
I don't think so, I've checked the sample and it all happens sequentially, each SNS message is sent individually. It might be more of an issue of sending 200 SNS messages that need to be rerouted to an SQS queue (per subscriber in the sample) while sending 200 messages to another SQS queue at the same time.
In the worst case in the sample, we are firing 8 times 200 messages to 8 different SQS queues. I'm not sure where the slowness comes from, could be from SNS creating a thread, a client and such per message sent, or in SQS.

**Comment by thrau** [comment]
I used the following script to try and reproduce the numbers:

```python
import os
import timeit

import boto3
from mypy_boto3_sns import SNSClient
from mypy_boto3_sqs import SQSClient

os.environ["AWS_REGION"] = "us-east-1"
os.environ["AWS_DEFAULT_REGION"] = "us-east-1"
os.environ["ACCESS_KEY_ID"] = "test"
os.environ["SECRET_ACCESS_KEY"] = "test"

sqs: SQSClient = boto3.client("sqs", endpoint_url="http://localhost:4566")
sns: SNSClient = boto3.client("sns", endpoint_url="http://localhost:4566")

topic_name = "my-topic"
topic_arn = None

num_subscribers = 10


def setup():
    global topic_arn

    result = sns.create_topic(Name=topic_name)
    topic_arn = result['TopicArn']

    for i in range(num_subscribers):
        queue_url = sqs.create_queue(QueueName=f"subscriber-queue-{i}")['QueueUrl']
        response = sqs.get_queue_attributes(QueueUrl=queue_url, AttributeNames=["QueueArn"])
        queue_arn = response["Attributes"]["QueueArn"]

        sns.subscribe(
            Protocol="sqs",
            TopicArn=topic_arn,
            Endpoint=queue_arn,
        )


def send_message():
    sns.publish(TopicArn=topic_arn, Message="my-message", Subject="my-subject")


def main():
    n = 200
    result = timeit.timeit(stmt=send_message, setup=setup, number=n)
    print(
        f"send message: {n} requests took {result:.4f}s = {(result * 1000) / n:.4f} ms/op = {n / result:.2f} req/sec"
    )


if __name__ == "__main__":
    main()
```

Here's what I got:


| num_subscribers | time_taken | time_per_op | req_per_sec |
| --------------- | ---------- | ----------- | ----------- |
| 0               | 0.4702s    | 2.3512 ms/op | 425.31      |
| 1               | 2.0520s    | 10.2599 ms/op | 97.47       |
| 2               | 3.3984s    | 16.9922 ms/op | 58.85       |
| 3               | 2.5259s    | 12.6294 ms/op | 79.18       |
| 4               | 3.3176s    | 16.5882 ms/op | 60.28       |
| ...              |                 |                           |                 |
| 10              | 3.6454s    | 18.2269 ms/op | 54.86       |


I can't really reproduce the depending behavior on the number of subscribers. it naturally takes a while for messages to arrive in the queue, but since we move the messages from sns to sqs asynchronously, the actual `SNS.Publish` request latency flattens pretty quickly. 

I agree though that generally the latency could be much better. I think we have the same fundamental problem as we had in #7342, where cloudwatch would make one invocation to `ThreadPoolExecutor.submit` per request, which turns out to be pretty slow. In SNS we are doing this here: https://github.com/localstack/localstack/blob/815c3993a3d1c734d3b6416ec1525b7311dfe1f0/localstack/services/sns/publisher.py#L1098-L1111

I don't see a way around this right now. A better approach would probably have been a simple queue-based event loop and then batch the publish calls.

/cc @bentsku @steffyP

**Comment by bentsku** [comment]
Hello @davy-oo, 

We've worked to improve our handling of concurrent requests hitting LocalStack.
Could you try pulling the latest image and starting LocalStack with `GATEWAY_SERVER=twisted` and see if you would get better results?

What was missing from our previous benchmarks were all the calls for your SQS subscribers polling the queues, which could be flooding the gateway and sometimes block, which would lead to the very long time to receive. 
This behavior should be resolved with the feature flag above. 

Thanks!

**Comment by HarshCasper** [comment]
Due to the forthcoming changes in our distribution and development model ([read more](https://blog.localstack.cloud/the-road-ahead-for-localstack/)), this issue tracker is no longer being maintained, and this issue is now being closed. If you'd like us to take a look or need further support, please reach out through one of our support channels:

-   💬 [GitHub Discussions](https://github.com/orgs/localstack/discussions)
-   📖 [Help & Support](https://docs.localstack.cloud/aws/getting-started/help-support/)
